### PR TITLE
Add a Override property to each DependencyItem

### PR DESCRIPTION
**Pull request by troydai**
Issue: #2744 
Reviewer: @davidfowl @abpiskunov 
CC: @victorhurdugaci @muratg @moozzyk @BrennanConroy 

This change add a new information to the __Dependencies__ message about if one dependency is overridden by different version of the same package from different source.

Here's a sample in which the Library1's `Newstonsoft.Json` dependencies is overridden.
```
{{
  "Framework": {
    "FrameworkName": "DNX,Version=v4.5.1",
    "FriendlyName": "DNX 4.5.1",
    "ShortName": "dnx451",
    "RedistListPath": "C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\Framework\\.NETFramework\\v4.5.1\\RedistList\\FrameworkList.xml"
  },
  "RootDependency": "Main",
  "Dependencies": {
    "Main": {
      "Name": "Main",
      "DisplayName": "Main",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Main\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Library1",
          "Version": "1.0.0",
          "Override": false
        },
        {
          "Name": "Library2",
          "Version": "1.0.0",
          "Override": false
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
    "Library1": {
      "Name": "Library1",
      "DisplayName": "Library1",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Library1\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Newtonsoft.Json",
          "Version": "6.0.8",
          "Override": true
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
    "Library2": {
      "Name": "Library2",
      "DisplayName": "Library2",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Library2\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Newtonsoft.Json",
          "Version": "6.0.8",
          "Override": false
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
// Remaining part of the message are omitted.
```

### CODE REVIEW

File: test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageTypes.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Comment (troydai) [inline test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageTypes.cs]: I'll add license here.

File: test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageTypes.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Dnx.DesignTimeHost.FunctionalTests.Util
+{
+    public class DthMessageTypes
```
Comment (troydai) [inline test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageTypes.cs]: I'll remove this. We need this eventually but I won't add it in this PR.

**Comment by abpiskunov** [comment]
Troy, ask was to get override=true if package does override other packages, not when its version is overriden... How do i know who's version does override?

Thanks,
Anton

On Oct 1, 2015, at 8:15 PM, Troy Dai <notifications@github.com<mailto:notifications@github.com>> wrote:


Issue: #2744<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fissues%2f2744&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=ryaUhV%2bcqNMzb79sEN5HHwYWkEaWGIL1BUHEDX1Qeqk%3d>
Reviewer: @davidfowl<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdavidfowl&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=lE6yn9RF%2fWw0VbYzLOIpNF6Ug0WXrEXvuR3Y4Gx0Pg0%3d> @abpiskunov<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fabpiskunov&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=vj9Svtc4bweKUFw2hNnr%2bWJqRvYi7t91pHbhE02Ox7s%3d>
CC: @victorhurdugaci<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fvictorhurdugaci&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=YgqKJXj0220F2chEP8BXJJ0151Ol9mgTu0%2be3qUBl8M%3d> @muratg<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fmuratg&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=4OdQ7IGybWILkYZIuC5%2b2AL9p9ibBGEnGYcVvCiIEyM%3d> @moozzyk<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fmoozzyk&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=M9Oj1lMSAiRrau8B3q%2bnJbF1A9ic%2fnRMkb026ylJMyE%3d> @BrennanConroy<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fBrennanConroy&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=J2cCW7k1S4lVDSZtVK6gbHcjDZDXjpAELGBkJR3cKkM%3d>

This change add a new information to the Dependencies message about if one dependency is overridden by different version of the same package from different source.

Here's a sample in which the Library1's Newstonsoft.Json dependencies is overridden.

{{
  "Framework": {
    "FrameworkName": "DNX,Version=v4.5.1",
    "FriendlyName": "DNX 4.5.1",
    "ShortName": "dnx451",
    "RedistListPath": "C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\Framework\\.NETFramework\\v4.5.1\\RedistList\\FrameworkList.xml"
  },
  "RootDependency": "Main",
  "Dependencies": {
    "Main": {
      "Name": "Main",
      "DisplayName": "Main",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Main\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Library1",
          "Version": "1.0.0",
          "Override": false
        },
        {
          "Name": "Library2",
          "Version": "1.0.0",
          "Override": false
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
    "Library1": {
      "Name": "Library1",
      "DisplayName": "Library1",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Library1\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Newtonsoft.Json",
          "Version": "6.0.8",
          "Override": true
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
    "Library2": {
      "Name": "Library2",
      "DisplayName": "Library2",
      "Version": "1.0.0",
      "Path": "C:\\code\\DNX\\artifacts\\TestOut\\DthTestServer.DthDependencies_HighestOverride\\dnx-clr-win-x86.1.0.0-dev\\src\\Library2\\project.json",
      "Type": "Project",
      "Resolved": true,
      "Dependencies": [
        {
          "Name": "Newtonsoft.Json",
          "Version": "6.0.8",
          "Override": false
        },
        {
          "Name": "fx/mscorlib",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/System.Core",
          "Version": "4.0.0",
          "Override": false
        },
        {
          "Name": "fx/Microsoft.CSharp",
          "Version": "4.0.0",
          "Override": false
        }
      ],
      "Errors": [],
      "Warnings": []
    },
// Remaining part of the message are omitted.


________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/aspnet/dnx/pull/2888<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=OjOiSx5ysqYjiRBDGcKmcKYvVR2NLsCtT%2f1zIHXjTlc%3d>

Commit Summary

  *   Add a Override property to each DependencyItem

File Changes

  *   A misc/DependencyOverrideNearest/global.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-0&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=KE6C6z%2fyzc53EqUBZiRN7NstMK5oKw3vLG1%2fqVbqU50%3d> (3)
  *   A misc/DependencyOverrideNearest/src/Library/project.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-1&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=6L6MNH7uJ9fCqEuBZvj4RiVxA3zUvsba9K%2bvNL1jqUs%3d> (8)
  *   A misc/DependencyOverrideNearest/src/Main/project.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-2&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=FbS901l3B0aXZWpV11EVsLDaa8BsU2FU0zlyFY9Wh2A%3d> (9)
  *   A misc/DependencyOverrideSibling/global.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-3&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=IITEPOdMRRJYUB2hEpHlBD76ll60C0%2b1eiKzMXH2rIk%3d> (3)
  *   A misc/DependencyOverrideSibling/src/Library1/project.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-4&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=9yTqW7Jy3ksnoeuUe0kH85BoM%2bRRgrJJpQ8loLwM4eA%3d> (8)
  *   A misc/DependencyOverrideSibling/src/Library2/project.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-5&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=2BnO35gDFT%2fpTCWho3z1OvVdd%2bTefyb9pbtYyoE5CjI%3d> (8)
  *   A misc/DependencyOverrideSibling/src/Main/project.json<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-6&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=OqAmEyMBC8PKZuarWu0YglrxfZuDV9pjgS0JcBHTSl0%3d> (9)
  *   M src/Microsoft.Dnx.DesignTimeHost/Models/OutgoingMessages/DependencyItem.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-7&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=qC1qnRZafvcmLy0BpjKe%2f9DKfAD51NQ%2bN3qnbNfIzTU%3d> (2)
  *   M src/Microsoft.Dnx.DesignTimeHost/ProjectStateResolver.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-8&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=xEBiS2PjsrbMxAsn7muzhPlPEjO92%2bM0DCckJwRip%2fQ%3d> (34)
  *   M test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/DthStartupTests.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-9&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=7ZG6yLk8B%2f25nSnEaFiF4x4%2bF4foKCoTGO%2fu2IFAUyM%3d> (92)
  *   M test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Infrastructure/DthTestClient.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-10&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=Nc89jlww8ya5sQmKwpn%2flMfR6Jrxqtznrc7NrM7CGqc%3d> (32)
  *   M test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageCollectionExtension.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-11&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=aBsgIobs3reoSqxkBF2j1qFhJe482b5wW2H9pTdftEA%3d> (24)
  *   A test/Microsoft.Dnx.DesignTimeHost.FunctionalTests/Util/DthMessageTypes.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888%2ffiles%23diff-12&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=9bcddXD8clWIvZ6isf7OtI9E6xB9qqk5GNuFjHyL%2fZs%3d> (14)

Patch Links:

  *   https://github.com/aspnet/dnx/pull/2888.patch<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888.patch&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=uRxpxvXf2M%2f6anpyURQgMBK2eknuBmjVjwaIuzaL4Do%3d>
  *   https://github.com/aspnet/dnx/pull/2888.diff<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2faspnet%2fdnx%2fpull%2f2888.diff&data=01%7c01%7canton.piskunov%40microsoft.com%7c9ccc3e98d70f434fbd1908d2cabe925c%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=KUIhp9Ssr25wgoscpW9jWr8i1beReKni0%2fPnk7SkUeY%3d>

-

**Comment by troydai** [comment]
@davidfowl correct me if I'm wrong.

@abpiskunov To find out which dependency wins is actually tricker than figure out who lost. If we purely analysis based on current lock file, it may leads into ambiguous situation. For example multiple libraries resolved to `Newtonsoft.Json\6.0.8` and they're all equally responsible under the rules then we can't deterministically choose one as the really winner. On the other hand a overridden dependency is much more clear.

To have absolute knowledge of which dependency is responsible of the overridden we need to go back to the restore process and really leave traces in the lock file, which to me is a bigger change in runtime and I think we need to carefully think about the trade off.

**Comment by troydai** [comment]
After examine the prototype we decide to use different approach for this problem.

**Comment by abpiskunov** [comment]
what is the new issue tracking it?

**Comment by troydai** [comment]
Please check the details in #2744. And I'm writing new issues for the new solution.

**Pull request closed without merging**